pkg/systemd/logs.js

```diff
@@ -192,7 +197,50 @@ $(function() {
             }
         }
 
+        function clear_service_list() {
+            $('#journal-services-list').empty()
+                    .append($('<li>').append($('<a>')
+                            .text(_("All"))
+                            .attr('data-service', "")))
+                    .append($('<li>')
+                            .addClass("divider"));
+        }
+
+        function load_service_filters(match, options) {
+            var service_options = Object.assign({ "output": "verbose" }, options);
+            var cmd = journal.build_cmd(match, service_options)[0].join(" ");
+            cmd += " | grep SYSLOG_IDENTIFIER= | sort -u";
+            cockpit.spawn(["bash", "-ec", cmd], { host: options.host, superuser: "try" })
+                    .then(function(entries) {
+                        entries.split("\n").forEach(function(entry) {
+                            if (entry)
+                                current_services.add(entry.substr(entry.indexOf('=') + 1));
```

Comment:
Both this and `prepend_services()` only ever add to `current_services`, it never gets reset. That seems like a bug? It's awkward that building this set happens in two different places, but that's a matter of optimization I figure (we can avoid the second run?) But neither of these two are conditional, so it seems they always both run?

Reply:
Ah, the fourth commit optimizes this. (It's not easy to follow, argh jQuery spaghetti code :-/). Can you please squash this?

Reply:
`prepend_services()` adds to the set because if there is a new log from a service that logged for the first time in this context than we want to add it to the list. (`prepend_services()` is used for only adding logs that happened after we queried all logs). Squashing of course is totally fine.

Reply:
Ah thanks, then I misunderstood this. Is `current_services` ever reset?

Reply:
Yes, before the last commit on every change of any filter including services one, with the last commit only when the services list is re-read from journal.